heron/common/src/java/com/twitter/heron/common/basics/WakeableLooper.java

```diff
@@ -44,22 +44,34 @@
   private final PriorityQueue<TimerTask> timers;
 
   // The tasks would be invoked before exit
-  private final ArrayList<Runnable> exitTasks;
+  private final List<Runnable> exitTasks;
 
   // For selector since there is bug in selector.select(timeout): we could not
   // use a timeout > 10 * Integer.MAX_VALUE
   // So here we set Integer.MAX_VALUE as the infinite future
   // We will also multiple 1000*1000 to convert mill-seconds to nano-seconds
   private static final Duration INFINITE_FUTURE = Duration.ofMillis(Integer.MAX_VALUE);
   private volatile boolean exitLoop;
+  private boolean lockAddNewTasks = false;
```

Comment:
If we want to use this lock, it needs to be an atomicBoolean, or needs to be volatile.

my concern is that simple boolean might have code cache and can be inconsistent between threads.